### PR TITLE
Expand on the provided example code

### PR DESCRIPTION
**Pull request by bigfootjon**
While I was working on another PR I thought it would be a good idea to have a local FBShipIt testbench that I could use to make sure I'm not breaking anything.

That turned out to be way harder than I anticipated. The provided code in the Example section of the README was a little out of date in some parts and very out of date in others. I spent several hours nuking my destination repo, stepping through things in the debugger, and reading ShipIt internals.

The thing that killed me is that in theory I know a little about how ShipIt works. I can't imagine the experience for a newcomer to the project.

So to save others my experience I updated the example code and documented how it works (see demo/README.md). I also specifically didn't add the `-example` filename suffix as I think it's a good idea to typecheck it to know when the example code we're providing breaks externally. (Sorry Yunus I know how you hate when the TravisCI builds break)

I imagine this will necessitate changing ShipIt's internal `getPathMappings` but that's not included in `fb-examples`, I wonder who's fault that is :P

I'd also like to consider adding some simple test invocation CI step like `demo/run_shipit.php --skip-destination-pull --skip-destination-init --skip-push` or something but one thing at a time.

### CODE REVIEW

File: demo/DemoGitHubUtils.php [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ * LICENSE file in the root directory of this source tree.
+ */
+
+namespace Facebook\ShipIt;
+
+final class DemoGitHubUtils extends ShipItGitHubUtils {
+
+  static string $committer_name = "FBShipIt Demo Committer";
+  static string $committer_user = 'CHANGEME';
+  static string $committer_email = "demo@example.com";
```
Comment (zpao) [inline demo/DemoGitHubUtils.php]: Sharing feedback from internal linting

```suggestion
  static string $committerName = "FBShipIt Demo Committer";
  static string $committerUser = 'CHANGEME';
  static string $committerEmail = "demo@example.com";
```

File: demo/DemoGitHubUtils.php [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+  public static function getCredentialsForProject(
```
Comment (zpao) [inline demo/DemoGitHubUtils.php]:
```suggestion
<<__Override>>
  public static function getCredentialsForProject(
```

File: demo/DemoSourceInitPhase.php [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        'write' => $_ ==> $this->skip(),
+      ),
+    ];
+  }
+
+  <<__Override>>
+  public function runImpl(ShipItBaseConfig $config): void {
+    $local_path = $config->getSourcePath();
+
+    $sh_lock = ShipItRepo::createSharedLockForPath($local_path);
```
Comment (zpao) [inline demo/DemoSourceInitPhase.php]: Want to release this lock?

File: demo/DemoSourceInitPhase.php [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+      ->setRetries(2)
+      ->setFailureHandler(
+        $_ ==> (
+          new ShipItShellCommand($local_parent_path, 'rm', '-rf', $local_path)
+        )->runSynchronously(),
+      )
+      ->runSynchronously();
+  }
+
+  public static function isMonorepo(string $name): bool {
```
Comment (zpao) [inline demo/DemoSourceInitPhase.php]: Unused
```suggestion
  public static function isMonorepo(string $_name): bool {
```

File: demo/run_importit.php [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  ShipItBaseConfig,
+  ShipItChangeset,
+  ShipItCleanPhase,
+  ShipItPullPhase,
+  ShipItGitHubInitPhase,
+  ShipItRepoSide,
+  ShipItTransport,
+};
+
+class ImportDemoProject {
```
Comment (zpao) [inline demo/run_importit.php]:
```suggestion
final class ImportDemoProject {
```

File: demo/run_importit.php [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+      |> ImportItPathFilters::moveDirectories(
+        $$,
+        ShipDemoProject::getPathMappings(),
+      );
+  }
+
+  public static function cliMain(): void {
+    // The repository state will be updated and modified, so we need to use a
+    // consistent destination repository for all ImportIt phases.
+    $source_repo_getter = (ShipItBaseConfig $c) ==> {
```
Comment (zpao) [inline demo/run_importit.php]: unused

File: demo/run_importit.php [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+          ),
+        ],
+      )
+    )
+      ->run();
+  }
+}
+
+<<__EntryPoint>>
+async function main(): Awaitable<void> {
```
Comment (zpao) [inline demo/run_importit.php]: Lint is saying `genMain` - not sure if that's fine?

File: demo/run_shipit.php [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+      ),
+      new ShipItPushPhase(),
+    ];
+
+    (new ShipItPhaseRunner($config, $phases))->run();
+  }
+}
+
+<<__EntryPoint>>
+async function main(): Awaitable<void> {
```
Comment (zpao) [inline demo/run_shipit.php]: Same comment about `gen`

File: demo/run_shipit.php [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+namespace Facebook\ShipIt;
+
+class ShipDemoProject {
```
Comment (zpao) [inline demo/run_shipit.php]: same comment about final

File: demo/DemoSourceInitPhase.php (hunk 3 above)
Comment (bigfootjon) [reply demo/DemoSourceInitPhase.php]: I just copied the behaviour from [FBRepoInitPhase.php-example](https://github.com/facebook/fbshipit/blob/master/fb-examples/lib/shipit/FBRepoInitPhase.php-example) ¯\_(ツ)_/¯

File: demo/run_importit.php (hunk 7 above)
Comment (bigfootjon) [reply demo/run_importit.php]: I'll write it as `mainAsync` and that should get [converted into `genMain` by ImportIt](https://github.com/facebook/fbshipit/blob/d64bc534be7c99f979f9ec758160e386913b2402/fb-examples/lib/config/FBWWWBasePlugin.php-example#L102)

**Pull request closed without merging**